### PR TITLE
[Performance] Update xgrammar-related constrained decoding

### PR DESCRIPTION
**Pull request by DarkSharpness**
## Modifications

Update the xgrammar-related constrained decoding part into a new API, including how to allocate `vocab_mask`, modify the `vocab_mask` and apply the `vocab_mask` to the `logits`.

In addition, we now use the custom cuda kernel provided by `xgrammar` to mask the `logits`, which brings a significant improvement in e2e performance (see [json_schema benchmark](https://github.com/sgl-project/sglang/tree/main/benchmark/json_schema))

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/constrained/outlines_backend.py [hunk 1]
```diff
@@ -81,10 +81,20 @@ def jump_and_retokenize(
...
         self.state = next_state
 
-    def fill_vocab_mask(self, vocab_mask: torch.Tensor):
+    def allocate_vocab_mask(
+        self, vocab_size: int, batch_size: int, device
+    ) -> torch.Tensor:
+        return torch.zeros(batch_size, vocab_size, dtype=torch.bool, device=device)
+
+    def fill_vocab_mask(self, vocab_mask: torch.Tensor, batch_id: int) -> None:
+        vocab_mask = vocab_mask[batch_id]
```
Comment (merrymercy) [inline python/sglang/srt/constrained/outlines_backend.py]:
```suggestion
    def fill_vocab_mask(self, vocab_mask: torch.Tensor, idx: int) -> None:
        vocab_mask = vocab_mask[idx]
```

File: python/sglang/srt/constrained/xgrammar_backend.py [hunk 2]
```diff
@@ -80,19 +85,24 @@ def jump_and_retokenize(
...
             assert self.matcher.accept_token(new_output_ids[i])
 
-    def fill_vocab_mask(self, vocab_mask: torch.Tensor):
+    def allocate_vocab_mask(
+        self, vocab_size: int, batch_size: int, device
+    ) -> torch.Tensor:
+        return self.matcher.allocate_token_bitmask(vocab_size, batch_size)
+
+    def fill_vocab_mask(self, vocab_mask: torch.Tensor, batch_id: int) -> None:
         # Note that this bitmask is a bitset, not bool
```
Comment (merrymercy) [inline python/sglang/srt/constrained/xgrammar_backend.py]:
```suggestion
    def fill_vocab_mask(self, vocab_mask: torch.Tensor, idx: int) -> None:
```

File: python/sglang/srt/constrained/xgrammar_backend.py [hunk 3]
```diff
@@ -80,19 +85,24 @@ def jump_and_retokenize(
...
-        bitmask = self.matcher.get_next_token_bitmask()
-        # Mask the tokens that are not allowed
-        vocab_mask[
-            self.matcher.get_rejected_tokens_from_bitmask(bitmask, self.vocab_size)
-        ] = 1
+        self.matcher.fill_next_token_bitmask(vocab_mask, batch_id)
```
Comment (merrymercy) [inline python/sglang/srt/constrained/xgrammar_backend.py]:
```suggestion
        self.matcher.fill_next_token_bitmask(vocab_mask, idx)
```
